atlasdb-timelock-server/src/main/java/com/palantir/atlasdb/timelock/paxos/PaxosServerImplementation.java

```diff
@@ -0,0 +1,233 @@
+import java.util.concurrent.Executors;
+import java.util.stream.Collectors;
+
+import javax.net.ssl.SSLSocketFactory;
+
+import com.google.common.base.Optional;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Sets;
+import com.google.common.net.HostAndPort;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import com.palantir.atlasdb.factory.Leaders;
+import com.palantir.atlasdb.http.AtlasDbHttpClients;
+import com.palantir.atlasdb.http.NotCurrentLeaderExceptionMapper;
+import com.palantir.atlasdb.timelock.ServerImplementation;
+import com.palantir.atlasdb.timelock.TimeLockServices;
+import com.palantir.atlasdb.timelock.config.PaxosConfiguration;
+import com.palantir.atlasdb.timelock.config.TimeLockServerConfiguration;
+import com.palantir.leader.LeaderElectionService;
+import com.palantir.leader.PaxosLeaderElectionService;
+import com.palantir.leader.PingableLeader;
+import com.palantir.leader.proxy.AwaitingLeadershipProxy;
+import com.palantir.lock.LockService;
+import com.palantir.lock.impl.LockServiceImpl;
+import com.palantir.paxos.PaxosAcceptor;
+import com.palantir.paxos.PaxosLearner;
+import com.palantir.paxos.PaxosProposer;
+import com.palantir.paxos.PaxosProposerImpl;
+import com.palantir.remoting.ssl.SslSocketFactories;
+import com.palantir.timestamp.PersistentTimestampService;
+import com.palantir.timestamp.TimestampService;
+
+import io.dropwizard.setup.Environment;
+
+public class PaxosServerImplementation implements ServerImplementation {
+    public static final String LEADER_NAMESPACE = "__leader";
+
+    private final PaxosConfiguration paxosConfiguration;
+    private final Environment environment;
+
+    private Set<String> remoteServers;
+    private Optional<SSLSocketFactory> optionalSecurity = Optional.absent();
+
+    private LeaderElectionService leaderElectionService;
+    private PaxosResource paxosResource;
+
+    public PaxosServerImplementation(PaxosConfiguration configuration, Environment environment) {
+        this.paxosConfiguration = configuration;
+        this.environment = environment;
+    }
+
+    @Override
+    public void onStartup(TimeLockServerConfiguration configuration) {
+        registerPaxosResource();
+
+        if (paxosConfiguration.sslConfiguration().isPresent()) {
+            optionalSecurity = constructOptionalSslSocketFactory(paxosConfiguration);
+        }
+
+        registerLeaderElectionService(configuration);
+        environment.jersey().register(new NotCurrentLeaderExceptionMapper());
+    }
+
+    private void registerPaxosResource() {
+        paxosResource = PaxosResource.create(paxosConfiguration.paxosDataDir().toString());
+        paxosResource.addClient(LEADER_NAMESPACE);
+        environment.jersey().register(paxosResource);
+    }
+
+    private void registerLeaderElectionService(TimeLockServerConfiguration configuration) {
+        remoteServers = getRemotePaths(configuration);
```

Comment:
A lot of code is reused in the `createPaxosBackedTimestampService` method below, too.